src/dbnode/storage/series/buffer.go

```diff
@@ -473,119 +440,346 @@ func (b *dbBuffer) FetchBlocksMetadata(
+			break
+		}
+	}
+	// Insert new time without allocating new slice.
+	b.inOrderBlockStarts = append(starts, timeZero)
+	// Update to new slice
+	starts = b.inOrderBlockStarts
+	copy(starts[idx+1:], starts[idx:])
+	starts[idx] = newTime
+}
 
+func (b *dbBuffer) inOrderBlockStartsRemove(removeTime time.Time) {
+	starts := b.inOrderBlockStarts
+	// There shouldn't be that many starts here, so just linear search through.
+	for i, t := range starts {
+		if t.Equal(removeTime) {
+			b.inOrderBlockStarts = append(starts[:i], starts[i+1:]...)
+			return
+		}
+	}
+}
+
+// BufferBucketVersions is a container for different versions (from
+// different flushes to disk) of buffer buckets in the database.
+type BufferBucketVersions struct {
+	buckets           []*BufferBucket
+	start             time.Time
+	opts              Options
+	lastReadUnixNanos int64
+	bucketPool        *BufferBucketPool
+}
+
+func (b *BufferBucketVersions) resetTo(
+	start time.Time,
+	opts Options,
+	bucketPool *BufferBucketPool,
+) {
+	// nil all elements so that they get GC'd.
+	for i := range b.buckets {
+		b.buckets[i] = nil
+	}
+	b.buckets = b.buckets[:0]
 	b.start = start
-	b.encoders = append(b.encoders, inOrderEncoder{
-		encoder: encoder,
-	})
-	b.bootstrapped = nil
+	b.opts = opts
 	atomic.StoreInt64(&b.lastReadUnixNanos, 0)
-	b.drained = false
+	b.bucketPool = bucketPool
 }
 
-func (b *dbBufferBucket) finalize() {
-	b.resetEncoders()
-	b.resetBootstrapped()
+func (b *BufferBucketVersions) streams(ctx context.Context, wType WriteType) []xio.BlockReader {
+	var res []xio.BlockReader
+	for _, bucket := range b.buckets {
+		if bucket.wType == wType {
+			res = append(res, bucket.streams(ctx)...)
+		}
+	}
+
+	return res
 }
 
-func (b *dbBufferBucket) empty() bool {
-	for _, block := range b.bootstrapped {
-		if block.Len() > 0 {
-			return false
+func (b *BufferBucketVersions) streamsLen() int {
+	res := 0
+	for _, bucket := range b.buckets {
+		res += bucket.streamsLen()
+	}
+	return res
+}
+
+func (b *BufferBucketVersions) write(
+	timestamp time.Time,
+	value float64,
+	unit xtime.Unit,
+	annotation []byte,
+	wType WriteType,
+) (bool, error) {
+	return b.writableBucketCreate(wType).write(timestamp, value, unit, annotation)
+}
+
+func (b *BufferBucketVersions) merge(wType WriteType) (int, error) {
+	res := 0
+	for _, bucket := range b.buckets {
+		// Only makes sense to merge buckets that are writable.
+		if bucket.version == writableBucketVer && wType == bucket.wType {
+			merges, err := bucket.merge()
+			if err != nil {
+				return res, nil
+			}
+			res += merges
 		}
 	}
-	for _, elem := range b.encoders {
-		if elem.encoder != nil && elem.encoder.NumEncoded() > 0 {
-			return false
+
+	return res, nil
+}
+
+func (b *BufferBucketVersions) removeBucketsUpToVersion(version int) {
+	// Avoid allocating a new backing array.
+	nonEvictedBuckets := b.buckets[:0]
+
+	for _, bucket := range b.buckets {
+		bVersion := bucket.version
+		// TODO(juchan): deal with ColdWrite too.
```

Comment:
Can you elaborate in this comment what at least might need to happen with ColdWrite when we do want to come back and "deal with ColdWrite"?